### PR TITLE
Add go decoder

### PR DESCRIPTION
**Pull request by pwaller**
First attempt at a go decoder.

I sort of wish I had done this before, since it highlights various ways that
the format is not ideal.

It makes heavy use of the `msgp` code which can be `go generate`'d, except
for Meta.FontInfos, which is mishandled and had to be manually fixed. I'll
file a bug upstream for that.

TODO:

- Also some code which invokes pdf2msgpack and parses that output.

### CODE REVIEW

**Comment by pwaller** [comment]
Example output:

```
$ pdf2msgpack --font-info 0-example.pdf | p2m-reader                                        
2016/07/15 16:40:36 FileName: "0-example.pdf"
2016/07/15 16:40:36 Pages: 4
2016/07/15 16:40:36 Creator: "Microsoft® Office Word 2007"
2016/07/15 16:40:36 Producer: "Microsoft® Office Word 2007"
p2m.FontInfo{Name:"ABCDEE+Cambria,Bold", Type:"TrueType", Encoding:"WinAnsi", Embedded:true, Subset:true, ToUnicode:false}
p2m.FontInfo{Name:"ABCDEE+Cambria", Type:"TrueType", Encoding:"WinAnsi", Embedded:true, Subset:true, ToUnicode:false}
p2m.FontInfo{Name:"ABCDEE+Calibri", Type:"TrueType", Encoding:"WinAnsi", Embedded:true, Subset:true, ToUnicode:false}
2016/07/15 16:40:36 Page i=1 s={595.4 841.8} len(paths)=4 len(glyphs)=34
2016/07/15 16:40:36 Page i=2 s={595.4 841.8} len(paths)=238 len(glyphs)=300
2016/07/15 16:40:36 Page i=3 s={595.4 841.8} len(paths)=237 len(glyphs)=301
2016/07/15 16:40:36 Page i=4 s={595.4 841.8} len(paths)=237 len(glyphs)=306

```

**Comment by pwaller** [comment]
I'm tempted to put this into another repository, to avoid having to pull in all of the vendor'd dependencies of pdf2msgpack if you just want to use the p2m go code. Thoughts?

Any thoughts @djui ?

**Comment by djui** [comment]
Not sure I understand. Could you elaborate on the motivation for this? I thought it works ok currently.

**Comment by pwaller** [comment]
Right now there is one in the pdftables.com tree which just parses metadata and nothing else. We have a complete one in Python.

My motivation for this is to write a more complete one in go, for the purposes of rapidly analysing many files, and opening the floor to algorithm experiments in go.

**Comment by djui** [comment]
Is this ready to review/merge?

**Comment by pwaller** [comment]
Misclick. Sorry. Yes, it is ready enough. There may be further fixes later but it does work now.

File: go/p2m-reader/main.go [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+
+	var p p2m.Page
+loop:
+	for {
+		err := r.Next(&p)
+		switch err {
+		default:
+			log.Fatal(err)
+		case io.EOF:
+			break loop
```
Comment (djui) [inline go/p2m-reader/main.go]: I guess for this specific implementation a return would be equivalent.

File: go/p2m-reader/main.go [hunk 2]
```diff
@@ -0,0 +1,44 @@
+package main
+
+import (
+	"io"
+	"log"
+	"os"
+
+	"github.com/kr/pretty"
```
Comment (djui) [inline go/p2m-reader/main.go]: Should these be vendored?

File: go/p2m-reader/main.go [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+
+	"github.com/scraperwiki/pdf2msgpack/go/p2m"
```
Comment (djui) [inline go/p2m-reader/main.go]: Should these be vendored?

File: go/p2m/decode.go [hunk 4]
```diff
@@ -0,0 +1,96 @@
+package p2m
+
+import "github.com/tinylib/msgp/msgp"
```
Comment (djui) [inline go/p2m/decode.go]: Should these be vendored?

**Comment by pwaller** [comment]
I consider this more of a library (+ a demo app) than an application, so: no, I don't think we should vendor anyhting.,

File: go/p2m/decode.go [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+		return
+	}
+	if sz != 2 {
+		err = msgp.ArrayError{Wanted: 2, Got: sz}
+		return
+	}
+
+	err = p.Type.DecodeMsg(dc)
+	if err != nil {
+		return err
```
Comment (djui) [inline go/p2m/decode.go]: Could be return without `err` for consistency.

**Comment by djui** [comment]
Alright, then I guess the options are like you mentioned: (i) pull it into its own repo or (ii) leave the dependencies explicitly unvendored. I agree with (ii).

I think the README should get an update and section explaing why it's here, how to build it and what it can do.

Then besides the minor things, lgtm, and I can merge this.

**Comment by pwaller** [comment]
:+1:

**Comment by djui** [comment]
- [ ] Update README or add README under `/go/README.md`.

**Pull request merged**